File: app/src/main/java/com/wavecat/inline/preferences/CheckBox.java
package com.wavecat.inline.preferences;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.view.View;

import com.google.android.material.checkbox.MaterialCheckBox;

import org.luaj.vm2.LuaValue;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;

@SuppressLint("ViewConstructor")
public class CheckBox extends MaterialCheckBox implements Preference {

    private String sharedKey;

    private LuaValue listener;

    private boolean defaultValue;

    @SuppressWarnings("unused")
    public CheckBox(Context context, String sharedKey, String text) {
        super(context);
        this.sharedKey = sharedKey;

        setText(text);
    }

    @SuppressWarnings("unused")
    public CheckBox(Context context, String text, LuaValue listener) {
        super(context);
        this.listener = listener;

        setText(text);
    }

    @SuppressWarnings("unused")
    public CheckBox(Context context, String text) {
        super(context);

        setText(text);
    }

    @SuppressWarnings("unused")
    public CheckBox setListener(LuaValue listener) {
        this.listener = listener;

        return this;
    }

    @SuppressWarnings("unused")
    public CheckBox setSharedKey(String sharedKey) {
        this.sharedKey = sharedKey;

        return this;
    }

    @SuppressWarnings("unused")
    public CheckBox setDefault(boolean defaultValue) {
        this.defaultValue = defaultValue;

        return this;
    }

    @SuppressWarnings("unused")
    public LuaValue getListener() {
        return listener;
    }

    @SuppressWarnings("unused")
    public String getSharedKey() {
        return sharedKey;
    }

    @Override
    public View getView(SharedPreferences preferences) {
        setOnCheckedChangeListener(null);

        if (sharedKey != null && preferences != null)
            setChecked(preferences.getBoolean(sharedKey, defaultValue));

        setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (sharedKey != null && preferences != null)
                preferences
                        .edit()
                        .putBoolean(sharedKey, isChecked)
                        .apply();

            if (listener != null)
                listener.call(LuaValue.valueOf(isChecked), CoerceJavaToLua.coerce(CheckBox.this));
        });

        return this;
    }
}

File: app/src/main/java/com/wavecat/inline/preferences/views/CheckBox.kt
@file:Suppress("unused", "MemberVisibilityCanBePrivate")

package com.wavecat.inline.preferences.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.view.View
import android.widget.CompoundButton
import androidx.core.content.edit
import com.google.android.material.checkbox.MaterialCheckBox
import com.wavecat.inline.preferences.Preference
import com.wavecat.inline.utils.dp
import org.luaj.vm2.LuaValue
import org.luaj.vm2.lib.jse.CoerceJavaToLua


/**
 * Custom checkbox implementation for preferences with Lua scripting support.
 *
 * Extends MaterialCheckBox to provide integration with SharedPreferences
 * for persistent storage and Lua callback functionality for dynamic behavior.
 * Supports both preference-backed and listener-based checkbox configurations.
 *
 * @see MaterialCheckBox
 * @see Preference
 * @see SharedPreferences
 */
@SuppressLint("ViewConstructor")
class CheckBox : MaterialCheckBox, Preference {

    var sharedKey: String? = null
        private set

    var listener: LuaValue? = null
        private set

    private var defaultValue = false

    constructor(context: Context?, sharedKey: String?, text: String?) : this(context) {
        this.sharedKey = sharedKey
        setText(text)
    }

    constructor(context: Context?, text: String?, listener: LuaValue?) : this(context) {
        this.listener = listener
        setText(text)
    }

    constructor(context: Context?, text: String?) : this(context) {
        setText(text)
    }

    constructor(context: Context?) : super(context) {
        minimumHeight = 36.dp
        translationX = (-6).dp.toFloat()
    }

    fun setListener(listener: LuaValue?): CheckBox {
        this.listener = listener
        return this
    }

    fun setSharedKey(sharedKey: String?): CheckBox {
        this.sharedKey = sharedKey
        return this
    }

    fun setDefault(defaultValue: Boolean): CheckBox {
        this.defaultValue = defaultValue
        return this
    }

    override fun getView(preferences: SharedPreferences?, requestFocus: () -> Unit): View {
        setOnCheckedChangeListener(null)

        sharedKey?.let {
            if (preferences != null)
                isChecked = preferences.getBoolean(sharedKey, defaultValue)
        }

        setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            sharedKey?.let {
                preferences?.edit {
                    putBoolean(it, isChecked)
                }
            }

            if (listener != null)
                listener!!.call(
                    LuaValue.valueOf(isChecked),
                    CoerceJavaToLua.coerce(this@CheckBox)
                )
        }

        return this
    }
}
